### PR TITLE
Optimize map down formula for non injective functions

### PR DESCRIPTION
**Pull request by mudathirmahgoub**
The following compares the previous implementation with the new one. 

image

### CODE REVIEW

File: src/theory/bags/inference_generator.cpp [hunk 1]
```diff
@@ -424,31 +424,16 @@ std::tuple<InferInfo, Node, Node> InferenceGenerator::mapDown(Node n, Node e)
   Node sum_zero = d_nm->mkNode(Kind::APPLY_UF, sum, d_zero);
   Node baseCase = d_nm->mkNode(Kind::EQUAL, sum_zero, d_zero);
 
-  // guess the size of the preimage of e
-  Node preImageSize =
-      d_sm->mkSkolemFunction(SkolemFunId::BAGS_MAP_PREIMAGE_SIZE, {f, A, e});
+  // guess the size of distinct elements in A
+  Node size =
+      d_sm->mkSkolemFunction(SkolemFunId::BAGS_MAP_PREIMAGE_SIZE, {f, A});
```
Comment (ajreynol) [inline src/theory/bags/inference_generator.cpp:429]: Can you update the documentation for the skolem for this? Also the ALF type rule will need to change (https://github.com/cvc5/cvc5/blob/main/proofs/alf/cvc5/theories/Bags.smt3#L36) to remove "U".

File: src/theory/bags/inference_generator.cpp [hunk 2]
```diff
@@ -415,7 +415,7 @@ std::tuple<InferInfo, Node, Node> InferenceGenerator::mapDown(Node n, Node e)
   Node f = n[0];
   Node A = n[1];
   // declare an uninterpreted function uf: Int -> T
-  Node uf = d_sm->mkSkolemFunction(SkolemFunId::BAGS_MAP_PREIMAGE, {f, A, e});
+  Node uf = d_sm->mkSkolemFunction(SkolemFunId::BAGS_MAP_PREIMAGE, {f, A});
```
Comment (ajreynol) [inline src/theory/bags/inference_generator.cpp:418]: Ditto for updating doc/ALF signature.

**Review by ajreynol — CHANGES_REQUESTED** [review]
What is the performance impact of this change? Is there any regression that can be added that is newly solved?

File: src/theory/bags/inference_generator.cpp (hunk 1 above)
Comment (mudathirmahgoub) [reply src/theory/bags/inference_generator.cpp]: Done.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**